README.md
# quicklm
A generative language model for low ressource data

### Example

```shell
complete("mi ku do") ------> "mi ku do ahwan kɔ fɔ xa kan"
```


checkpoints/history.txt
BigramLanguageModel_model_at_560_L1_33: {'bs': 64, 'blsz': 256, 'max_iters': 3370, 'lr': 0.0003, 'n_embd': 384, 'n_head': 6, 'n_layer': 6, 'drop': 0.2}
BigramLanguageModel_model_at_565_L1_31: {'bs': 64, 'blsz': 256, 'max_iters': 3370, 'lr': 0.0003, 'n_embd': 384, 'n_head': 6, 'n_layer': 6, 'drop': 0.2}
GPT_model_at_3020_L1_35: {'bs': 64, 'blsz': 256, 'gamma': 0.1, 'step_size': 2, 'lr_step_size': 1500, 'max_iters': 4980, 'lr': 0.0003, 'n_embd': 384, 'n_head': 6, 'n_layer': 6, 'drop': 0.1, 'ffwd_dim': None} 


asr.py
# @WIP
"""Here we will train an automatic fon recognition model,
based on our generative language model. We will use
some huggingface resources (https://huggingface.co/) """

import pickle
import pandas as pd
import torch
from torch import nn
from torch import optim
from torch.utils.data import Dataset, DataLoader
from transformers import (AutoModelForCTC, AutoTokenizer,
                          AutoConfig)
from datasets import DatasetDict, Dataset as TDataset
from utils import get_processing_data
import evaluate

torch.manual_seed(4224)
checkpoint_path = "checkpoints/"
MODEL_CKPT = "facebook/wav2vec2-large-xlsr-53"
model_name = "wav2vec2-large-xlsr-53"
SAMPLING_RATE = 16000
BATCH_SIZE = 8
learning_rate = 1e-4
iteration = 0
EPOCHS = 20

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
torch.cuda.empty_cache()

# Prepare dataset
with open('data/audio_train.pkl', 'rb') as file:
    train = pickle.load(file)
    train['waveform'] = train['waveform'].values

with open('data/audio_test.pkl', 'rb') as file:
    test = pickle.load(file)
    test['waveform'] = test['waveform'].values

train_dict = TDataset.from_pandas(train)
test_dict = TDataset.from_pandas(test)

data = DatasetDict({"train": train_dict, "test": test_dict})
_, vocab_size = get_processing_data(data)

data = pd.concat([train, test], axis=0, ignore_index=True)

# we only focus on waveform and text columns
data = data[['text', 'waveform']]

max_waveform_len = data['waveform'].apply(len).max()
max_target_len = data['text'].apply(len).max()

cut = int(len(data) * 0.9)
train, test = data.iloc[:cut, :], data.iloc[cut:, :]
train.reset_index(drop=True, inplace=True), test.reset_index(drop=True, inplace=True)


def _config():
    config = AutoConfig.from_pretrained(MODEL_CKPT)

    tokenizer_type = config.model_type if config.tokenizer_class is None else None
    config = config if config.tokenizer_class is not None else None

    return config, tokenizer_type


def _get_tokenizer():
    config, tokenizer_type = _config()
    tokenizer = AutoTokenizer.from_pretrained(
        f"{MODEL_CKPT.split('/')[-1]}/",
        config=config,
        tokenizer_type=tokenizer_type,
        unk_token="[UNK]",
        pad_token="[PAD]",
        word_delimiter_token="|",
        truncation=True,
    )
    return tokenizer


class SpeechDataset(Dataset):
    def __init__(self, data):
        super().__init__()
        self.data = data
        self.max_waveform_len = max_waveform_len
        self.max_target_len = max_target_len
        self.tokenizer = _get_tokenizer()

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        waveform = torch.tensor(self.data['waveform'][idx])
        transcription = self.data['text'][idx]
        waveform_len = waveform.shape[0]

        if waveform_len < self.max_waveform_len:
            # add padding to normalize waveforms length
            padding_len = self.max_waveform_len - waveform_len
            padding = torch.zeros((1, padding_len)).squeeze(0)
            waveform = torch.cat([waveform, padding], dim=0)
        elif waveform_len > self.max_waveform_len:
            raise Exception("Exceed waveform length")

        target = self.tokenizer(transcription, return_tensors="pt").input_ids.squeeze(0)

        target_len = target.shape[0]
        if target_len < self.max_target_len:
            # Padding the transcription
            padding_len = self.max_target_len - target_len
            padding = torch.tensor([29] * padding_len, dtype=torch.long)
            target = torch.cat([target, padding], dim=0)
        elif target_len > self.max_target_len:
            raise Exception("Exceed target length")

        assert len(waveform) == self.max_waveform_len and len(target) == self.max_target_len
        return waveform, target


dataset, test_dataset = SpeechDataset(train), SpeechDataset(test)
dataloader = DataLoader(dataset, batch_size=BATCH_SIZE, shuffle=True)
test_dataloader = DataLoader(test_dataset, batch_size=1, shuffle=True)

# Train

model = AutoModelForCTC.from_pretrained(MODEL_CKPT)
tokenizer = _get_tokenizer()


class SpeechToTextModel(nn.Module):
    def __init__(self, model, tokenizer):
        super().__init__()
        self.model = model
        self.tokenizer = tokenizer
        self.net = nn.Sequential(
            nn.Linear(32, 16),
            nn.ReLU(),
            nn.Linear(16, vocab_size),
            nn.Dropout(0.2),
        )

    def forward(self, input_ids, attention_mask):
        logits = self.model(input_ids, attention_mask).logits
        # fine-tuning architecture
        logits = self.net(logits)
        return logits.log_softmax(2)

    def transcribe(self, waveform):
        # waveform normalisation
        # features = torch.log(waveform + 1e-9)
        # input_ids = self.tokenizer(features, return_tensors='pt').input_ids

        # transcription
        logits = self.forward(input_ids=waveform, attention_mask=None)
        predicted_ids = torch.argmax(logits, dim=-1)
        transcription = self.tokenizer.batch_decode(predicted_ids)[0]

        return transcription


model = SpeechToTextModel(model=model, tokenizer=tokenizer)
optimizer = optim.AdamW(model.parameters(), lr=learning_rate)
loss_fn = nn.CTCLoss()

model.to(device)
for epoch in range(EPOCHS):
    model.train()
    for idx, batch in enumerate(dataloader):
        optimizer.zero_grad()
        input_ids, targets = batch
        logits = model(input_ids=input_ids.to(device), attention_mask=None)
        logits = logits.transpose(0, 1)

        T, B, C = logits.shape  # input_length, batch size, number of class
        input_lengths = torch.full(size=(B,), fill_value=T, dtype=torch.long)
        target_lengths = torch.randint(low=0, high=vocab_size - 1, size=(B,), dtype=torch.long)
        loss = loss_fn(logits, targets.to(device), input_lengths=input_lengths, target_lengths=target_lengths)

        print(f"Loss at epoch {epoch} batch {idx}: {loss.item()}")
        loss.backward()
        optimizer.step()
    saving_info = {
        "bs": BATCH_SIZE,
        "epoch": epoch + iteration,
        "lr": learning_rate,
    }

    # For test purpose
    model.eval()
    for waveform_test_w, waveform_test_t in test_dataloader:
        transcription = model.transcribe(waveform_test_w)
        print("trans: ", transcription, "\n", "actual: ", tokenizer.batch_decode(waveform_test_t)[0])
        break

    torch.save({
        'epoch': epoch,
        'model_state_dict': model.state_dict(),
        'optimizer_state_dict': optimizer.state_dict(),
        'loss': loss.item(),
    }, f"{checkpoint_path}/{model_name}_{epoch}.pt")
    open(f'{checkpoint_path}/asr_history.txt', 'a', encoding="utf-8").write(
        f"{model_name}_{epoch}: {saving_info} \n"
    )


utils.py
import json
import re


def remove_special_characters(batch):
    """ Remove punctuation and lower the text """
    chars_to_ignore_regex = '[\,\?\.\!\-\;\:\"]'
    batch["text"] = re.sub(chars_to_ignore_regex, '', batch["text"]).lower() + " "
    return batch


def extract_all_chars(batch):
    """ Get all unique characters in text """
    all_text = " ".join(batch["text"])
    vocab = list(set(all_text))
    return {"vocab": [vocab], "all_text": [all_text]}


def get_processing_data(dataset):
    dataset = dataset.map(remove_special_characters)

    vocabs = dataset.map(
        extract_all_chars,
        batched=True,
        batch_size=-1,
        keep_in_memory=True,
        remove_columns=dataset.column_names["train"]
    )
    vocab_list = list(set(vocabs["train"]["vocab"][0]) |
                      set(vocabs["test"]["vocab"][0])
                      )
    vocab_dict = {v: k for k, v in
                  enumerate(sorted(vocab_list))}  # {'a': 1, 'b': 2, etc}

    vocab_dict["|"] = vocab_dict[" "]
    del vocab_dict[" "]
    vocab_dict["[UNK]"] = len(vocab_dict)
    vocab_dict["[PAD]"] = len(vocab_dict)

    with open(f'wav2vec2-large-xlsr-53/vocab.json', 'w', encoding="utf-8") as vocab_file:
        json.dump(vocab_dict, vocab_file)

    return dataset, len(vocab_dict)


predict.py
import torch
import json
from MicroGPT import (NGramLanguageModel, GPT,
                      encode, decode, device)

model_name = "BigramLanguageModel_model_at_565_L1_31"


def get_config():
    with open("checkpoints/history.txt", "r") as file:
        history = file.read().splitlines()
        for h in history:
            spl = h.split(':')
            name, config = spl[0], str(':'.join(spl[1:])).replace('\'', '\"')
            if name == model_name:
                print("Config found in history...")
                return json.loads(config)


def generate(write=False):
    model = NGramLanguageModel()

    checkpoint = torch.load(f"checkpoints/{model_name}.pt")
    model.load_state_dict(checkpoint['model_state_dict'])
    model.eval()
    print(torch.tensor(encode("mi ku do"), dtype=torch.long, device=device).unsqueeze(0))
    output = decode(model.generate(torch.tensor(encode("mi ku do"), dtype=torch.long, device=device).unsqueeze(0),
                                   max_gen_tokens=20)[0].tolist())
    print(output)
    if write:
        open('generate.txt', 'w', encoding="utf-8").write(output)


if __name__ == '__main__':
    generate()


MicroGPT.py
import os
import math
import torch
import torch.distributed as dist
import torch.multiprocessing as mp
import torch.nn as nn
from torch.nn import functional as F
from torch.nn.parallel import DistributedDataParallel as DDP
from dataclasses import dataclass
import time

# hyper parameters
torch.manual_seed(4224)
batch_size = 64
block_size = 256  # maximum context to look at for the next prediction
max_iters = 3000
eval_interval = 500
learning_rate = 3e-4
step_size = 2
lr_step_size = max_iters // step_size if max_iters > 2 * step_size else 10
gamma = 0.1
device = torch.device("cpu" if torch.cuda.is_available() else "cpu")
eval_iters = 200
n_embed = 384
n_head = 6
n_layer = 6
dropout = 0.2
ddp = False # distributed training

from_checkpoint = True
iteration = 0
checkpoint_path = "checkpoints/"
checkpoint_interval = 20

torch.cuda.empty_cache()

with open("data/fongbe.txt", "r", encoding="utf-8") as f:
    text = f.read()
# text = text[:1500000]
# get a list of characters in the text
chars = sorted(list(set(text)))
vocab_size = len(chars)
print(f"Char numbers {len(text)}\nvocab size {vocab_size}")

# tokenization
# str -----> int
# we create a mapping ffrom characters to integers
ch2idx = {ch: idx for idx, ch in enumerate(chars)}
idx2ch = {idx: ch for idx, ch in enumerate(chars)}

encode = lambda s: [ch2idx[ch] for ch in s]  # the encoder take a string and return the list of integers
decode = lambda idx_l: ''.join([idx2ch[idx] for idx in idx_l])  # take a list of integers and return the string

# we could use tiktoken (open ai) or sentence piece (google) tokenization

# Convert text encode to torch.Tensor

data = torch.tensor(encode(text), dtype=torch.long)

# split the dataset to train and val set: 90% and 10%
n = int(0.9 * len(data))
train_data = data[:n]
val_data = data[n:]


def gelu(x):
    """
    Implementation of the GELU activation from open ai repo
    Reference: https://github.com/openai/gpt-2/blob/master/src/model.py#L25
    """
    return 0.5 * x * (1.0 + torch.tanh(math.sqrt(2.0 / math.pi) * (x + 0.044715 * torch.pow(x, 3.0))))


def new_softmax(x, dim=-1):
    """
    Softmax used for gpt-2
    """
    x = x - torch.max(x, dim=dim, keepdim=True)[0]
    ex = torch.exp(x)
    return ex / torch.sum(ex, dim=dim, keepdim=True)


"""
In order to train our data, we should chunk data, to gain in computation and 
efficiency (chunking and batching). Take data by block
"""


def get_batch(split):
    data = train_data if split == "train" else val_data

    ix = torch.randint(len(data) - block_size, (batch_size,))  # we take k random value with size of bloc_size
    x = torch.stack([data[i:i + block_size] for i in ix])
    y = torch.stack([data[i + 1:i + block_size + 1] for i in ix])

    return x.to(device), y.to(device)


@torch.no_grad()
def estimate_loss(model):
    out = {}
    model.eval()
    for split in ['train', 'val']:
        losses = torch.zeros(eval_iters)
        for k in range(eval_iters):
            X, y = get_batch(split)
            logits = model(X)

            # make this piece of code as func
            B, T, C = logits.shape
            logits = logits.view(B * T, C)
            y = y.view(B * T)
            loss = F.cross_entropy(logits, y)
            # ------#----------

            losses[k] = loss.item()
        out[split] = losses.mean()
    model.train()
    return out


class Head(nn.Module):
    """One head of self attention"""

    def __init__(self, head_size):
        super().__init__()
        self.query = nn.Linear(n_embed, head_size, bias=False)  # what i'm looking for ?
        self.key = nn.Linear(n_embed, head_size, bias=False)  # what do i contains
        self.value = nn.Linear(n_embed, head_size, bias=False)  # element that can be aggregate
        self.register_buffer('tril', torch.tril(torch.ones(block_size, block_size)))

        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        B, T, C = x.shape

        k = self.key(x)  # (B, T, head_size)
        q = self.query(x)  # (B, T, head_size)
        w = q @ k.transpose(-2, -1) * C ** -0.5  # (B, T, head_size) @ (B, head_size, T) = (B,T,T)

        w = w.masked_fill(self.tril[:T, :T] == 0, float('-inf'))
        w = F.softmax(w, dim=-1)
        w = self.dropout(w)
        v = self.value(x)  # (B,T,C)
        out = w @ v  # (B,T,T) @ (B,T,C) => (B,T,C)

        return out


class MultiHeadAttention(nn.Module):
    def __init__(self, num_heads, head_size):
        super().__init__()
        self.heads = nn.ModuleList([Head(head_size) for _ in range(num_heads)])
        self.proj = nn.Linear(n_embed, n_embed)  # linear projection of outcome
        self.dropout = nn.Dropout(dropout)

    def forward(self, x):
        out = torch.cat([h(x) for h in self.heads], dim=-1)
        out = self.dropout(self.proj(out))
        return out


# The feed forward net, data obtain after attention need to think individually
class FeedForward(nn.Module):
    def __init__(self, n_embed, h_dim=None, dropout=dropout):
        super().__init__()
        h_dim = 4 * n_embed if h_dim is None else h_dim  # the size of ffwd layer
        self.ln1 = nn.Linear(n_embed, h_dim)
        self.ln2 = nn.Linear(h_dim, n_embed)
        self.drop = nn.Dropout(dropout)

        # in order to replicate the last checkpoint
        # self.net = nn.Sequential(
        #     nn.Linear(n_embed, h_dim),
        #     nn.ReLU(),
        #     nn.Linear(h_dim, n_embed),
        #     nn.Dropout(dropout),
        # )

    def forward(self, x):
        x = self.ln1(x)
        x = gelu(x)
        x = self.ln2(x)
        x = self.drop(x)
        return x


# Create transformer block
class TransformerBlock(nn.Module):
    """ Transformer block"""

    def __init__(self, n_embed, n_head, ffwd_config=None):
        super().__init__()
        head_size = n_embed // n_head
        self.sa_head = MultiHeadAttention(n_head, head_size)
        if ffwd_config:
            self.ffwd = FeedForward(n_embed, **ffwd_config)
        else:
            self.ffwd = FeedForward(n_embed)
        self.ln1 = nn.LayerNorm(n_embed)
        self.ln2 = nn.LayerNorm(n_embed)

    def forward(self, x):
        # applying residual connection https://arxiv.org/pdf/1512.03385.pdf
        x = x + self.sa_head(self.ln1(x))
        x = x + self.ffwd(self.ln2(x))
        return x


@dataclass
class GPTConfig:
    vocab_size: int = 50257  # total number of tokens
    n_embed: int = 1024
    n_head: int = 16
    n_hid: int = 4096
    n_layer: int = 48
    ffwd_dim: int = 8192
    dropout: float = 0.1


@torch.no_grad()
def generate_token(model, idx, max_gen_tokens):
    """Generate new tokens from the current context idx (B, T): the last character
    Not taking history"""
    for _ in range(max_gen_tokens):
        idx_temp = idx if idx.size(1) \
                          <= block_size else idx[:, -block_size:]  # take out from the idx the last
        # block_size token
        logits = model(idx_temp)
        logits = logits[:, -1, :]  # take the last embeddings in each batch (B, C)
        probs = F.softmax(logits, dim=-1)  # get probabilities by applying softmax
        # then get the next idx
        idx_next = torch.multinomial(probs, num_samples=1)  # (B, 1)
        idx = torch.cat((idx, idx_next), dim=1)

    return idx


class GPT(nn.Module):
    def __init__(self, vocab_size, n_embed, n_head, ffwd_dim, n_layer, dropout=0.1):
        super().__init__()
        self.ffwd_dim = ffwd_dim
        self.dropout = dropout
        self.token_embedding_table = nn.Embedding(vocab_size, n_embed)
        self.position_embedding_table = nn.Embedding(block_size, n_embed)
        self.drop = nn.Dropout(dropout)

        self.transformer_blocks = nn.ModuleList(
            [TransformerBlock(n_embed, n_head, ffwd_config={'h_dim': ffwd_dim, 'dropout': dropout}) for _ in
             range(n_layer)])
        self.ln_f = nn.LayerNorm(n_embed)
        self.lm_head = nn.Linear(n_embed, vocab_size)

        # apply weight tying https://paperswithcode.com/method/weight-tying
        self.token_embedding_table.weight = self.lm_head.weight

    def forward(self, idx):
        B, T = idx.shape
        tok_emb = self.token_embedding_table(idx)
        pos_emb = self.position_embedding_table(torch.arange(T, device=device))  # (T, C)
        x = self.drop(tok_emb + pos_emb)  # (B, T, C)
        # apply  attention (B,T,C)
        for transformer in self.transformer_blocks:
            x = transformer(x)

        x = self.ln_f(x)  # final layer normalisation (B,T,C)
        logits = self.lm_head(x)  # (B,T,vocab_size); @TODO: optimize this for inference time
        return logits

    def generate(self, idx, max_gen_tokens):
        return generate_token(self, idx, max_gen_tokens)


class NGramLanguageModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.token_embedding_table = nn.Embedding(vocab_size, n_embed)
        self.position_embedding_table = nn.Embedding(block_size, n_embed)
        self.blocks = nn.Sequential(*[TransformerBlock(n_embed, n_head) for _ in range(n_layer)])
        self.ln_f = nn.LayerNorm(n_embed)
        self.lm_head = nn.Linear(n_embed, vocab_size)

    def forward(self, idx):
        # n_embed = C
        B, T = idx.shape
        tok_emb = self.token_embedding_table(idx)  # (B, T) => (B, T, C) : batch, time to ..., chanel
        pos_emb = self.position_embedding_table(torch.arange(T, device=device))  # (T, C)
        x = tok_emb + pos_emb  # (B, T, C)
        x = self.blocks(x)  # apply  attention (B,T,C)
        x = self.ln_f(x)  # final layer normalisation (B,T,C)
        logits = self.lm_head(x)  # (B,T,vocab_size)

        # B, T, C = logits.shape
        # logits = logits.view(B * T, C)
        # ----up--- we are not compute loss here anymore
        return logits

    def generate(self, idx, max_gen_tokens):
        """Generate new tokens from the current context idx (B, T): the last character
        Not taking history"""
        return generate_token(self, idx, max_gen_tokens)


def training_loop(model, distributed=False, rank=None):
    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)
    scheduler = torch.optim.lr_scheduler.StepLR(optimizer=optimizer, step_size=lr_step_size, gamma=gamma)
    for iter in range(iteration, max_iters+iteration):
        print(f"Iteration {iter}")

        if iter % eval_interval == 0:
            losses = estimate_loss(model)
            print(f"step {iter}: train loss: {losses['train']:.4f} val loss: {losses['val']:.4f}")

        # sample batch
        xb, yb = get_batch('train')  # but how we ensure all data is learned ?

        if distributed:
            logits = model(xb.to(rank))
            yb = yb.to(rank)
        else:
            logits = model(xb)

        B, T, C = logits.shape
        logits = logits.view(B * T, C)
        yb = yb.view(B * T)
        optimizer.zero_grad(set_to_none=True)
        loss = F.cross_entropy(logits, yb)
        loss.backward()
        optimizer.step()
        scheduler.step()

        if iter % checkpoint_interval == 0:
            model_name = str(model.__class__.__name__)
            saving_info = {
                "bs": batch_size,
                "blsz": block_size,
                "gamma": gamma,
                "step_size": step_size,
                "lr_step_size": lr_step_size,
                "max_iters": max_iters + iteration,
                "lr": learning_rate,
                "n_embd": n_embed,
                "n_head": n_head,
                "n_layer": n_layer,
                "drop": dropout
            }
            if model_name == "GPT":
                saving_info['ffwd_dim'] = model.ffwd_dim
                saving_info['drop'] = model.dropout

            file_name = f"model_at_{iter}_L{str(float('{:.2f}'.format(loss))).replace('.', '_')}"
            torch.save({
                'iteration': iter,
                'model_state_dict': model.state_dict(),
                'optimizer_state_dict': optimizer.state_dict(),
                'loss': loss,
            }, f"{checkpoint_path}/{model_name}_{file_name}.pt")
            open(f'{checkpoint_path}/history.txt', 'a', encoding="utf-8").write(
                f"{model_name}_{file_name}: {saving_info} \n"
            )


if from_checkpoint:
    CKPT_PATH = 'checkpoints/GPT_model_at_3020_L1_35.pt'

    model = GPT(vocab_size, n_embed, n_head, None, n_layer, 0.1)

    checkpoint = torch.load(CKPT_PATH)
    model.load_state_dict(checkpoint['model_state_dict'])
    iteration = checkpoint['iteration']
    loss = checkpoint['loss']
    print(model, iteration, loss)

else:
    model = GPT(vocab_size, n_embed, n_head, None, n_layer, 0.1)
    model = model.to(device)


def distributed_training(rank, world_size):
    # create the default group
    dist.init_process_group("gloo", rank=rank, world_size=world_size)
    # model = model.to(rank)
    ddp_model = DDP(model) # device_ids=[rank]
    training_loop(model=ddp_model, distributed=True, rank=rank)
    dist.destroy_process_group()


if __name__ == '__main__':
    print(sum(p.numel() for p in model.parameters()) / 1e6, 'M parameters')

    if ddp:
        os.environ["MASTER_ADDR"] = "localhost"
        os.environ["MASTER_PORT"] = "29500"
        world_size = 4
        try:
            mp.spawn(distributed_training,
                     args=(world_size,), nprocs=world_size,
                     join=True)
        except Exception as e:
            raise e
    else:
        training_loop(model=model, distributed=False)

    p_tokens = model.generate(torch.zeros((1, 1), dtype=torch.long, device=device), max_gen_tokens=100)[0].tolist()

    for i in range(0, len(p_tokens), 3):
        print(decode(p_tokens[:i]))
        time.sleep(0.5)